node/src/chain/stall_chain_fsm.rs

```diff
@@ -0,0 +1,302 @@
+            state: State::Running,
+            recovery_blocks: BTreeMap::new(),
+            tip: Default::default(),
+            latest_finalized,
+            acc,
+        };
+
+        sm.update_tip(tip.inner().clone());
+        sm
+    }
+
+    /// Handles block received event
+    ///
+    /// Returns the new state of the FSM after processing the block
+    pub(crate) async fn on_block_received(&mut self, blk: &Block) -> State {
+        trace!(
+            event = "chain.block_received",
+            hash = to_str(&blk.header().hash),
+            height = blk.header().height,
+            iter = blk.header().iteration,
+        );
+
+        let tip = self.acc.read().await.get_curr_tip().await;
+
+        if self.tip.0.header().hash != tip.inner().header().hash {
+            // Tip has changed, which means a new block is accepted either due
+            // to normal block acceptance or fallback execution
+            self.recovery_blocks.clear();
+
+            if tip.is_final() {
+                self.latest_finalized = tip.inner().clone();
+            }
+
+            self.update_tip(tip.inner().clone());
+            self.state_transition(State::Running);
+        }
+
+        let curr = self.state;
+        match curr {
+            State::Running => self.on_running().await,
+            State::Stalled => {
+                if let Err(err) = self.on_stalled(blk).await {
+                    error!("Error while processing block: {:?}", err);
+                }
+            }
+            State::StalledOnFork => warn!("Stalled on fork"),
+        };
+
+        self.state
+    }
+
+    /// Returns recovery blocks as a vector
+    pub(crate) fn recovery_blocks(&self) -> Vec<Block> {
+        self.recovery_blocks.values().cloned().collect()
+    }
+
+    /// Handles a running state
+    async fn on_running(&mut self) {
+        if self.tip.1 + STALLED_TIMEOUT
+            < SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs()
+        {
+            // While we are still receiving blocks, no block
+            // has been accepted for a long time (tip has not changed
+            // recently)
+            self.on_accept_block_timeout().await
+        }
+    }
+
+    /// Handles block from wire in the `Stalled` state
+    async fn on_stalled(&mut self, new_blk: &Block) -> anyhow::Result<()> {
+        let key = new_blk.header().height;
+        self.recovery_blocks
+            .entry(key)
+            .or_insert_with(|| new_blk.clone());
+
+        info!(
+            event = "chain.recovery_block_added",
+            hash = to_str(&new_blk.header().hash),
+            height = new_blk.header().height,
+            iter = new_blk.header().iteration,
+        );
+
+        // Ensure all blocks from local_final until current_tip are
+        // collected
+        let from = self.latest_finalized.header().height + 1;
+        let to = self.tip.0.header().height + 1;
```

Comment:
Another thing: if in the future we filter out "equivalent" messages (e.g. blocks we already know), and we have a fork starting at a higher block than our last final, we might not collect "all" recovery blocks (because we would discard those we already have). So we would never trigger the revert+sync.

I think we should really just check if we receive a valid fork block, instead of waiting for the recover blocks list to be "complete"